File: companions/benchmarks/snapshot_executor_driver.py
from __future__ import annotations

import argparse
import json
import os
import urllib.error
import urllib.request
from pathlib import Path
from urllib.parse import urlparse


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--request", type=Path, required=True)
    parser.add_argument("--result", type=Path, required=True)
    parser.add_argument("--source", required=True)
    parser.add_argument("--revision", required=True)
    args = parser.parse_args()
    endpoint = os.getenv("ELLIOTT_BENCHMARK_EXECUTOR_ENDPOINT", "")
    token = os.getenv("ELLIOTT_BENCHMARK_EXECUTOR_TOKEN", "")
    parsed = urlparse(endpoint)
    if parsed.scheme not in {"http", "https"} or parsed.hostname not in {
        "127.0.0.1",
        "::1",
        "localhost",
    }:
        raise SystemExit("benchmark executor must be an HTTP loopback endpoint")
    if (
        parsed.username is not None
        or parsed.password is not None
        or parsed.query
        or parsed.fragment
    ):
        raise SystemExit("benchmark executor endpoint is invalid")
    if not token:
        raise SystemExit("benchmark executor token is required")
    operation = json.loads(args.request.read_text(encoding="utf-8"))
    payload = json.dumps(
        {
            "operation": operation,
            "driverSource": args.source,
            "driverRevision": args.revision,
        },
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")
    request = urllib.request.Request(
        endpoint.rstrip("/") + "/v1/benchmark",
        data=payload,
        method="POST",
        headers={
            "authorization": f"Bearer {token}",
            "content-type": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(
            request,
            timeout=operation["timeoutMilliseconds"] / 1000,
        ) as response:
            result = response.read(32 * 1024 * 1024 + 1)
    except urllib.error.URLError as error:
        raise SystemExit(f"benchmark executor failed: {error}") from error
    if len(result) > 32 * 1024 * 1024:
        raise SystemExit("benchmark executor result is too large")
    value = json.loads(result)
    args.result.write_text(
        json.dumps(value, separators=(",", ":"), sort_keys=True),
        encoding="utf-8",
    )
    args.result.chmod(0o600)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: companions/benchmarks/snapshot-executor-driver.ts
import { chmod, readFile, writeFile } from "node:fs/promises";
import { HTTP_INTERNAL_SERVER_ERROR } from "../../src/runtime/http";
import {
  canonicalJson,
  MAX_RESPONSE_BYTES,
  requireLoopbackEndpoint,
  wireError,
} from "../typescript/wire";

const OWNER_ONLY_FILE_MODE = 0o600;

const argument = (name: string): string => {
  const index = Bun.argv.indexOf(name);
  const value = index === -1 ? undefined : Bun.argv[index + 1];
  return value === undefined || value.length === 0
    ? wireError(`${name} is required`, HTTP_INTERNAL_SERVER_ERROR)
    : value;
};

const main = async (): Promise<void> => {
  const requestPath = argument("--request");
  const resultPath = argument("--result");
  const source = argument("--source");
  const revision = argument("--revision");
  const operation: unknown = JSON.parse(await readFile(requestPath, "utf8"));
  const executor = requireLoopbackEndpoint(
    Bun.env["ELLIOTT_BENCHMARK_EXECUTOR_ENDPOINT"],
    Bun.env["ELLIOTT_BENCHMARK_EXECUTOR_TOKEN"],
    "benchmark executor",
  );
  const response = await fetch(
    new URL("/v1/benchmark", executor.endpoint),
    {
      method: "POST",
      headers: {
        authorization: `Bearer ${executor.token}`,
        "content-type": "application/json",
      },
      body: canonicalJson({
        operation,
        driverSource: source,
        driverRevision: revision,
      }),
    },
  );
  if (!response.ok) {
    return wireError(
      `benchmark executor returned HTTP ${response.status}`,
      HTTP_INTERNAL_SERVER_ERROR,
    );
  }
  const encoded = await response.text();
  if (Buffer.byteLength(encoded) > MAX_RESPONSE_BYTES) {
    return wireError(
      "benchmark executor result is too large",
      HTTP_INTERNAL_SERVER_ERROR,
    );
  }
  JSON.parse(encoded);
  await writeFile(resultPath, encoded, { mode: OWNER_ONLY_FILE_MODE });
  await chmod(resultPath, OWNER_ONLY_FILE_MODE);
};

await main();
